### PR TITLE
Refactor Leaderboard Submission Eval Code.

### PR DESCRIPTION
**Pull request by alexzhang13**
## Description

The old leaderboard submission eval code expects a `metric()` function from the reference code specified by the leaderboard creator. Instead, we only expect the leaderboard creator to have 2 functions: the reference kernel itself, and an input data generator. We now handle checking implementations and runtimes on our end to avoid hacking and to allow us to quickly fix leaderboard issues. This change was per last Tuesday's meeting.

We will probably update `leaderboard_eval.py` over time, this is just a simple implementation to start that is usable for Python / Triton.

Example of what new reference code may look like for softmax:
```
import torch
from typing import List

def ref_kernel(x: torch.Tensor, dim: int = -1) -> torch.Tensor:
    """
    Reference implementation of the Softmax function using PyTorch's predefined functions.
    Args:
        x (torch.Tensor): Input tensor.
        dim (int): Dimension along which to apply softmax.
    Returns:
        torch.Tensor: Tensor after applying Softmax.
    """
    return torch.nn.functional.softmax(x, dim=dim)


def generate_input(seed: int = None, to_cuda: bool = True) -> List[torch.Tensor]:
    """
    Generates random input tensor of the specified shape.
    Args:
        seed (int): Random seed for reproducibility.
    Returns:
        torch.Tensor: Randomly generated tensor.
    """
    shapes = [(128, 64), (256, 64), (512, 64)]
    device = torch.device("cuda" if to_cuda else "cpu")

    if seed is not None:
        torch.manual_seed(seed)

    tensors = []
    for shape in shapes:
        tensors.append(torch.randn(shape, device=device))

    return tensors
```

## Checklist

Before submitting this PR, ensure the following steps have been completed:

- [x] Run the slash command `/verifyruns` on your own server.
  - Run the cluster bot on your server:
    ```bash
    python discord-bot.py
    ```
  - Start training runs with the slash command `/verifyruns`.
  - Verify that the bot eventually responds with:
    ```
    ✅ All runs completed successfully!
    ```
    (It may take a few minutes for all runs to finish. In particular, the GitHub
    runs may take a little longer. The Modal run is typically quick.)
  For more information on running a cluster bot on your own server, see
  README.md.

### CODE REVIEW

File: src/discord-cluster-manager/leaderboard_eval.py [hunk 1]
```diff
@@ -3,10 +3,56 @@
...
+import time
+from reference import ref_kernel, generate_input
+from train import custom_kernel
+
+
+def check_implementation():
+    for _ in range(10):  # check multiple times
+        input_tensors = generate_input()
+        for input in input_tensors:
+            custom_output = custom_kernel(input, dim=-1)
```
[the remark this answers is not in the document]
Comment (alexzhang13) [reply src/discord-cluster-manager/leaderboard_eval.py:16]: Good catch, for some reason I thought I removed these!

File: src/discord-cluster-manager/leaderboard_eval.py [hunk 2]
```diff
@@ -3,10 +3,56 @@
...
+            ref_output = ref_kernel(input, dim=-1)
+
+            if not torch.allclose(custom_output, ref_output, atol=1e-5):
+                print('mismatch found! custom implementation doesnt match reference.')
```
[the remark this answers is not in the document]
Comment (alexzhang13) [reply src/discord-cluster-manager/leaderboard_eval.py:20]: For now I've just wrapped it in an assert, we can change this later.

**Review by S1ro1 — APPROVED** [review]
lgtm
